### ISSUE TITLE
Replicate the C++ `git-p4` work in Go

### ISSUE THREAD

**Issue by dan-mckean** [sourcegraph/sourcegraph#25436]
**TL;DR:** [Customer](https://github.com/sourcegraph/accounts/issues/6716) has replicated `git-p4` in C++, this work is to port it to Go.

## History:
[This](https://docs.google.com/document/d/11x5BY3pyFIUD7LO1QWbz2L_ep7DO-X-K4nA5ZgrZoxA/edit) was the original summary of the problem.

@asdine did awesome work collaborating with [Customer](https://github.com/sourcegraph/accounts/issues/6716) to try and improve the performance of Perforce git-p4.

As it happens, the work we were able to do in Python (in close collaboration with [Customer](https://github.com/sourcegraph/accounts/issues/6716) themselves) was beaten by a C++ version that another of the [Customer](https://github.com/sourcegraph/accounts/issues/6716) engineers developed. 

> We tested it by trying to perform a full perforce history clone on a moderately large depot path (around 3390 CLs) from our Perforce server. Details follow.
> Total changelist count: ~3390
> Hardware: Sourcegraph VM with 72 GB RAM, 24 cores
> 
>Timings:
> - Original git-p4.py - 42 mins
> - Sourcegraph modified git-p4.py script with 24 threads (as of 20th August, with multithreading) - 2:45 mins
> - Sourcegraph modified git-p4.py script with 24 threads and print batches as of this week - 1:45 mins
> - Custom C++ client, with a naive algorithm, single threaded, as of 20th August: 84 mins
> - Custom C++ client with a smarter algorithm which matches the multithreaded git-p4.py script’s algorithm, 400 threads, as of yesterday - 26 seconds \o/
> 
>That makes the custom client ~100x faster than the original git-p4.py script and gets around 75% time reduction from the 1:45 mins metric. 

Some more info [about the approach we took and the results](https://docs.google.com/spreadsheets/d/1Cn7QjCFQLI6uH8WyHkpdeth16fA_wKkfGfsV6pww4sw/edit#gid=0).

## Detail:
[Here's the link to the C++ code](https://drive.google.com/file/d/12lSV6hZIPE9dqoZkkBbTXhXWNPW04PY9/view?usp=sharing)

There was a meeting to talk through how the C++ code works - [recording](https://docs.google.com/document/d/1KlIooLGiiR-2HB5C34YgUuM_VVGSoUuHLid7PzTweUo/edit?pli=1) and there is an [accompanying doc](https://docs.google.com/document/d/1_DRAGLrGfwBax-uaSiwUUnHyM9SA6mCZbzrXgAIHU4A/edit#heading=h.xlwupp5ccjj6).

**Comment by dan-mckean** [comment]
c++ client for perforce is under version control now at https://github.com/sourcegraph/p4-cpp-client

**Comment by dan-mckean** [comment]
We've no intention to do this at the moment.
